File: queueDraft.cpp
#include<iostream>
#include<queue>

using namespace std;

struct TreeNode {
  int val;
  TreeNode* left;
  TreeNode* right;
  TreeNode(): val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x): val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode* left, TreeNode* right): val(x), left(left), right(right) {}
};

class Solution {
public:
  bool recur(TreeNode* left, TreeNode* right) {
    if (left == NULL && right == NULL)
      return true;
    if (left == NULL || right == NULL || left->val != right->val)
      return false;
    return (recur(left->left, right->right) && recur(left->right, right->left));
  }
  bool isSymmetric(TreeNode* root) {
    return recur(root, root);
  }
};

class Solution1 {
public:
  bool isSymmetric(TreeNode* root) {
    if (root == NULL)
      return true;
    if (!root->left && !root->right)
      return true;
    if (!root-> left || !root->right)
      return false;

    queue<TreeNode*> queue;
    queue.push(root);
    queue.push(root);

    TreeNode* left, * right;

    while (!queue.empty())
    {
      left = queue.front();
      queue.pop();

      right = queue.front();
      queue.pop();

      if (left->val != right->val)
        return false;
      if (left->left && right->right)
      {
        queue.push(left->left);
        queue.push(right->right);
      }

      else if (left->left || right->right)
        return false;
      if (left->right && right->left)
      {
        queue.push(left->right);
        queue.push(right->left);
      }
      else if (left->right || right->left)
        return false;
    }
    return true;
  }
};

int main(int argc, char* argv[]) {
  TreeNode* pval3 = new TreeNode(3);
  TreeNode* pval2 = new TreeNode(3);
  TreeNode* pval1 = new TreeNode(1, pval2, pval3);

  Solution sol;
  Solution1 sol1;
  bool res = sol.isSymmetric(pval1);
  bool res1 = sol1.isSymmetric(pval1);

    cout << res << endl;
    cout << res1 << endl;

  delete pval3;
  delete pval2;
  delete pval1;
}












// end

File: queueDraft.py

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
      self.val = val
      self.left = left
      self.right = right


class Solution:
     def isSymmetric(self, root):
         print('hi')
         def recur(left, right):
             # base cases
             if (left == None and right == None):
                 return True
             if (left == None or right == None):
                 return False
             if (left.val != right.val):
                 return False
             return (recur(left.left, right.right) and recur(left.right, right.left))
         print('root', root)
         return recur(root, root)

s = Solution()
pval3 = TreeNode(2)
pval2 = TreeNode(2)
pval1 = TreeNode(1, pval2, pval3)
print('pval1', pval1)
res = s.isSymmetric(pval1)
print(res)













# end
